### PR TITLE
HDDS-6295. EC: Fix unaligned stripe write failure due to length overflow.

### PR DESCRIPTION
**Pull request by guihecheng**
Co-authored-by: Uma Maheswara Rao G <umagangumalla@cloudera.com>

## What changes were proposed in this pull request?

Fix unaligned stripe write failure due to length overflow and
eliminate the unnecessary variable `currentBlockGroupLen`.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-6295

## How was this patch tested?

existing UT in CI: https://github.com/guihecheng/ozone/actions/runs/1827271876
manual test:
./bin/ozone sh bucket create vol1/bucket1 --layout=FILE_SYSTEM_OPTIMIZED --replication=rs-10-4-1024k --type EC
./bin/ozone sh key put /vol1/bucket1/hosts /etc/hosts

a new ut suggested by Uma, tested before and after apply patch.

### CODE REVIEW

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/ECKeyOutputStream.java [hunk 1]
```diff
@@ -587,8 +578,17 @@ public static void padBufferToLimit(ByteBuffer buf, int limit) {
     buf.position(limit);
   }
 
-  private boolean isPartialStripe() {
-    return currentBlockGroupLen % (numDataBlks * ecChunkSize) > 0;
+  private boolean isPartialStripe(int stripeSize) {
+    return stripeSize > 0 && stripeSize < numDataBlks * ecChunkSize;
+  }
+
+  private int getCurrentStripeSize() {
```
Comment (umamaheswararao) [inline hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/ECKeyOutputStream.java:585]: Could you please rename this to getCurrentDataStripeSize?
Generally we refer stripe size as combination of data+parity. We we need only data.

**Comment by umamaheswararao** [comment]
I really want a test for this.
Below is the test code reproduces it. Please include this and update. Thanks a lot for your help @guihecheng 
```
  @Test
  public void test10D4PConfigWithPartialStripe()
      throws IOException {
    conf.set(OzoneConfigKeys.OZONE_SCM_BLOCK_SIZE,
        OzoneConfigKeys.OZONE_SCM_BLOCK_SIZE_DEFAULT);
    int dataBlks = 10;
    int parityBlk = 4;
    store.createVolume(volumeName);
    OzoneVolume volume = store.getVolume(volumeName);
    volume.createBucket(bucketName);
    OzoneBucket bucket = volume.getBucket(bucketName);

    try (OzoneOutputStream out = bucket.createKey(keyName, chunkSize,
        new ECReplicationConfig(dataBlks, parityBlk,
            ECReplicationConfig.EcCodec.RS, chunkSize), new HashMap<>())) {
      out.write(new byte[1023]);
    }
  }
```

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/ECKeyOutputStream.java (hunk 1 above)
Comment (guihecheng) [reply hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/ECKeyOutputStream.java:585]: Sure, thanks.

**Review by umamaheswararao — APPROVED** [review]
LGTM. Pending CI.

**Pull request merged**